### PR TITLE
Enable full functionality on SDCC

### PR DESCRIPTION
**Pull request by bkleiner**
Builds on #13

This PR adds full functionality on SDCC, without removing any features.

There are quite a lot of changes and i am open to splitting it into smaller chunks.

code-size for the freestyle vtx (worst case):
SDCC: 58132, 88.7% usage
Keil: 51960, 79,2% usage

Commit 51df352aa4dcd5baf38ba4e69db0062eaf445245 modifies the camera init and video detection to be more friendly to sdcc's memory management. review and further testing would be appreciated

Commit a8e3eb460450fd858906b4f3e19e58f5c34e3858 and 493c4633b2ac82cfe8afbf557179734df8f576f0 prevent a stack corruption in `DM6300_Init` which is present on Keil too, however does not case an issues with the original code. Nonetheless a fix should be like the commits mentioned should be implemented.

Commit e01ec8b0657e368e54536ab2ec5b041020ba639a and e16c1e18c1c713199e55ba3f4feaaebd8ae98e91 rework the DM6300 initialization code to replay the SPI register values from arrays instead of writing them out one by one. This produces most of the code-size reduction and allows us to fit the code so comfortably.
These commits also make the aforementioned stack corrupt rear its ugly head, both on keil and sdcc. 

I have tested the code with both compilers on my Whoop VTX and Foxeer camera, as this hardware combination only covers a portion of the code, more brave testers would be appreciated.

Many thanks to those that have contributed:
@pjpei for the initial hardware exploration and most of the timing and endian code
@ligenxxxx for his help on discord.
@SumolX for figuring out the memory corruption issue.

### CODE REVIEW

File: src/msp_displayport.c [hunk 1]
```diff
@@ -1629,115 +1633,99 @@ void set_vtx_param()
 }
 
 #ifdef INIT_VTX_TABLE
-void InitVtxTable()
-{
-    uint8_t i,j;
+static const uint8_t bf_vtx_band_table[6][31] = {
```
Comment (RoarFPV) [inline src/msp_displayport.c:1636]: This same type of conversion should be applied to `DM6300_init7` looks like it's inlined i2c commands

Comment (bkleiner) [reply src/msp_displayport.c:1636]: i dont quite follow, could you elaborate on that?
like split the regmap and compose in-function?
or set a ptr and run the `DM6300_write_reg_map` function on the very end?

Comment (RoarFPV) [reply src/msp_displayport.c:1636]: I see why you are confused, please ignore my comment, I failed to actually load the diff for dm6300.c and will now think about my life choices lol

**Comment by ligenxxxx** [comment]
[vtx_freestyle.zip](https://github.com/hd-zero/hdzero-vtx/files/9136793/vtx_freestyle.zip)
[vtx_whoop_lite.zip](https://github.com/hd-zero/hdzero-vtx/files/9136794/vtx_whoop_lite.zip)
[vtx_whoop.zip](https://github.com/hd-zero/hdzero-vtx/files/9136795/vtx_whoop.zip)
[vtx_race.zip](https://github.com/hd-zero/hdzero-vtx/files/9136796/vtx_race.zip)

**Comment by ligenxxxx** [comment]
[vtx_whoop_lite.zip](https://github.com/hd-zero/hdzero-vtx/files/9136798/vtx_whoop_lite.zip)

[vtx_whoop.zip](https://github.com/hd-zero/hdzero-vtx/files/9136799/vtx_whoop.zip)

[vtx_race.zip](https://github.com/hd-zero/hdzero-vtx/files/9136800/vtx_race.zip)

[vtx_freestyle.zip](https://github.com/hd-zero/hdzero-vtx/files/9136801/vtx_freestyle.zip)

**Comment by ligenxxxx** [comment]
I did a simple test, using whoop vtx, and found the following problems:
- When using HDZero nano v2, vdo disappears every few seconds. There is no problem with vdo when using foxeer camera.
- When using foxeer camera, OSD update speed seems unstable and looks very slow.

**Comment by bkleiner** [comment]
@ligenxxxx could you kindly give it another try?
i think i fixed the issues you raised, but i am not sure if the SD_3016 mode is functioning as it supposed to.

**Comment by ligenxxxx** [comment]
Build with SDCC:
[vtx_freestyle.zip](https://github.com/hd-zero/hdzero-vtx/files/9165968/vtx_freestyle.zip)
[vtx_race.zip](https://github.com/hd-zero/hdzero-vtx/files/9165969/vtx_race.zip)
[vtx_whoop.zip](https://github.com/hd-zero/hdzero-vtx/files/9165970/vtx_whoop.zip)
[vtx_whoop_lite.zip](https://github.com/hd-zero/hdzero-vtx/files/9165971/vtx_whoop_lite.zip)

I did the following test:
- VTX uses race v1
- VRX uses Bytefrost VRX and VRX4
- The VTX has been powered on many times, and occasionally fails to start. It is speculated that the reason is that the temperature of the VTX is too high, which can be solved by the cooling fan.
- VRX checked OSD refresh speed and VLQ, no problem.

**Comment by bkleiner** [comment]
yeah i have been experiencing intermittent boot failures (usually resulting in a black screen) _without_ a fan too.
The chip would be blistering hot, too hot to touch, ever since i added a fan to my setup however, it has been fine.
We might wanna look into that, but its seems it is not related to this PR and/or SDCC.

I gonna convert this to full PR and see if we can get some more brave testers before we merge?

**Pull request merged**